File: src/main/java/de/twometer/amongus/net/server/TaskGenerator.java
package de.twometer.amongus.net.server;

import de.twometer.amongus.model.Location;
import de.twometer.amongus.model.PlayerTask;
import de.twometer.amongus.model.TaskStage;
import de.twometer.amongus.model.TaskType;
import de.twometer.amongus.util.RandomUtil;

import java.util.List;

import static de.twometer.amongus.util.RandomUtil.getRandomItem;

public class TaskGenerator {

    private static final List<Location> WIRING_LOCATIONS = List.of(Location.Cafeteria, Location.Electrical, Location.Navigation, Location.Security, Location.Storage);

    private static final List<Location> ACCEPT_POWER_LOCATIONS = List.of(Location.Navigation, Location.Security, Location.Reactor, Location.Weapons, Location.O2, Location.Comms);

    private static final List<Location> DOWNLOAD_LOCATIONS = List.of(Location.Comms, Location.Navigation, Location.Electrical, Location.Cafeteria, Location.Weapons);

    private static final List<Location> GARBAGE_LOCATIONS = List.of(Location.Storage, Location.O2, Location.Cafeteria);

    private static final List<TaskType> COMMON_TASKS = List.of(
            TaskType.SwipeCard,
            TaskType.FixWiring
    );

    private static final List<TaskType> SHORT_TASKS = List.of(
            TaskType.EngineMgmt,
            TaskType.ChartCourse,
            TaskType.DivertPower,
            TaskType.StabilizeSteering,
            TaskType.PrimeShields,
            TaskType.UnlockManifold,
            TaskType.DataTransfer
    );

    private static final List<TaskType> LONG_TASKS = List.of(
            TaskType.Scan,
            TaskType.StartReactor,
            TaskType.InspectSamples,
            TaskType.ClearGarbage,
            TaskType.Shoot
    );

    public static PlayerTask newCommonTask() {
        return expand(getRandomItem(COMMON_TASKS));
    }

    public static PlayerTask newShortTask() {
        return expand(getRandomItem(SHORT_TASKS));
    }

    public static PlayerTask newLongTask() {
        return expand(getRandomItem(LONG_TASKS));
    }

    private static PlayerTask expand(TaskType taskType) {
        switch (taskType) {
            case SwipeCard:
                return new PlayerTask.Builder()
                        .addStage(new TaskStage(Location.Admin, TaskType.SwipeCard))
                        .build();
            case FixWiring:
                var builder = new PlayerTask.Builder();
                var locations = RandomUtil.getUniqueRandom(WIRING_LOCATIONS, 3);
                for (var loc : locations)
                    builder.addStage(new TaskStage(loc, TaskType.FixWiring));
                return builder.build();
            case EngineMgmt:
                return new PlayerTask.Builder()
                        .addStage(Location.UpperEngine, TaskType.EngineMgmt)
                        .addStage(Location.LowerEngine, TaskType.EngineMgmt)
                        .build();
            case ChartCourse:
                return makeSingleStageTask(Location.Navigation, TaskType.ChartCourse);
            case DivertPower:
                return new PlayerTask.Builder()
                        .addStage(Location.Electrical, TaskType.DivertPower)
                        .addStage(getRandomItem(ACCEPT_POWER_LOCATIONS), TaskType.AcceptPower)
                        .build();
            case StabilizeSteering:
                return makeSingleStageTask(Location.Navigation, TaskType.StabilizeSteering);
            case PrimeShields:
                return makeSingleStageTask(Location.Shields, TaskType.PrimeShields);
            case UnlockManifold:
                return makeSingleStageTask(Location.Reactor, TaskType.UnlockManifold);
            case DataTransfer:
                return new PlayerTask.Builder()
                        .addStage(getRandomItem(DOWNLOAD_LOCATIONS), TaskType.DataTransfer)
                        .addStage(Location.Admin, TaskType.DataTransfer)
                        .build();
            case Scan:
                return makeSingleStageTask(Location.MedBay, TaskType.Scan);
            case StartReactor:
                return makeSingleStageTask(Location.Reactor, TaskType.StartReactor);
            case InspectSamples:
                return makeSingleStageTask(Location.MedBay, TaskType.InspectSamples);
            case ClearGarbage:
                var garbageLocations = RandomUtil.getUniqueRandom(GARBAGE_LOCATIONS, 2).toArray(new Location[0]);
                return new PlayerTask.Builder()
                        .addStage(garbageLocations[0], TaskType.ClearGarbage)
                        .addStage(garbageLocations[1], TaskType.ClearGarbage)
                        .build();
            case Shoot:
                return makeSingleStageTask(Location.Weapons, TaskType.Shoot);
        }

        throw new IllegalArgumentException("Unknown task type " + taskType);
    }

    private static PlayerTask makeSingleStageTask(Location loc, TaskType type) {
        return new PlayerTask.Builder()
                .addStage(loc, type)
                .build();
    }

}

File: src/main/kotlin/de/twometer/amogus/server/TaskGenerator.kt
package de.twometer.amogus.server

import de.twometer.amogus.model.Location
import de.twometer.amogus.model.PlayerTask
import de.twometer.amogus.model.TaskStage
import de.twometer.amogus.model.TaskType
import de.twometer.amogus.model.TaskType.*
import kotlin.math.floor

object TaskGenerator {

    private val WIRING_LOCATIONS: List<Location> = listOf(
        Location.Cafeteria,
        Location.Electrical,
        Location.Navigation,
        Location.Security,
        Location.Storage
    )

    private val ACCEPT_POWER_LOCATIONS: List<Location> = listOf(
        Location.Navigation,
        Location.Security,
        Location.Reactor,
        Location.Weapons,
        Location.O2,
        Location.Comms
    )

    private val DOWNLOAD_LOCATIONS: List<Location> = listOf(
        Location.Comms,
        Location.Navigation,
        Location.Electrical,
        Location.Cafeteria,
        Location.Weapons
    )

    private val GARBAGE_LOCATIONS: List<Location> =
        listOf(Location.Storage, Location.O2, Location.Cafeteria)

    private val COMMON_TASKS: List<TaskType> = listOf(
        SwipeCard,
        FixWiring
    )

    private val SHORT_TASKS: List<TaskType> = listOf(
        EngineMgmt,
        ChartCourse,
        DivertPower,
        StabilizeSteering,
        PrimeShields,
        UnlockManifold,
        DataTransfer
    )

    private val LONG_TASKS: List<TaskType> = listOf(
        Scan,
        StartReactor,
        InspectSamples,
        ClearGarbage,
        Shoot
    )

    fun newCommonTask(): PlayerTask {
        return expand(COMMON_TASKS.getRandomItem())
    }

    fun newShortTask(): PlayerTask {
        return expand(SHORT_TASKS.getRandomItem())
    }

    fun newLongTask(): PlayerTask {
        return expand(LONG_TASKS.getRandomItem())
    }

    private fun expand(taskType: TaskType): PlayerTask {
        when (taskType) {
            SwipeCard -> return PlayerTask()
                .addStage(Location.Admin, SwipeCard)

            FixWiring -> {
                val builder = PlayerTask()
                val locations = getUniqueRandom(WIRING_LOCATIONS, 3)
                for (loc in locations) builder.addStage(loc, FixWiring)
                return builder
            }
            EngineMgmt -> return PlayerTask()
                .addStage(Location.UpperEngine, EngineMgmt)
                .addStage(Location.LowerEngine, EngineMgmt)

            ChartCourse -> return makeSingleStageTask(Location.Navigation, ChartCourse)
            DivertPower -> return PlayerTask()
                .addStage(Location.Electrical, DivertPower)
                .addStage(ACCEPT_POWER_LOCATIONS.getRandomItem(), AcceptPower)

            StabilizeSteering -> return makeSingleStageTask(
                Location.Navigation,
                StabilizeSteering
            )
            PrimeShields -> return makeSingleStageTask(Location.Shields, PrimeShields)
            UnlockManifold -> return makeSingleStageTask(Location.Reactor, UnlockManifold)
            DataTransfer -> return PlayerTask()
                .addStage(DOWNLOAD_LOCATIONS.getRandomItem(), DataTransfer)
                .addStage(Location.Admin, DataTransfer)

            Scan -> return makeSingleStageTask(Location.MedBay, Scan)
            StartReactor -> return makeSingleStageTask(Location.Reactor, StartReactor)
            InspectSamples -> return makeSingleStageTask(Location.MedBay, InspectSamples)
            ClearGarbage -> {
                val garbageLocations = getUniqueRandom(GARBAGE_LOCATIONS, 2).toTypedArray()
                return PlayerTask()
                    .addStage(garbageLocations[0], ClearGarbage)
                    .addStage(garbageLocations[1], ClearGarbage)

            }
            Shoot -> return makeSingleStageTask(Location.Weapons, Shoot)
            else -> throw IllegalArgumentException("Unknown task type $taskType")
        }
    }

    private fun makeSingleStageTask(loc: Location, type: TaskType): PlayerTask {
        return PlayerTask().addStage(loc, type)
    }

    private fun PlayerTask.addStage(loc: Location, type: TaskType): PlayerTask {
        this.stages.add(TaskStage(loc, type))
        return this
    }

    private fun <T> List<T>.getRandomItem(): T {
        return this[floor(Math.random() * size).toInt()]
    }

    private fun <T> getUniqueRandom(list: List<T>, num: Int): Set<T> {
        val set = HashSet<T>()
        do set.add(list.getRandomItem())
        while (set.size < num)
        return set
    }


}